src/ng/directive/ngOptions.js

```diff
@@ -325,12 +325,28 @@ var ngOptionsDirective = ['$compile', '$parse', function($compile, $parse) {
         // The option values were already computed in the `getWatchables` fn,
         // which must have been called to trigger `getOptions`
         var optionValues = valuesFn(scope) || [];
+        var optionValuesKeys;
 
-        var keys = Object.keys(optionValues);
-        keys.forEach(function getOption(key) {
+
+        if (!keyName && isArrayLike(optionValues)) {
```

Comment:
This check ensures that we don't treat array-like objects as array-like if we are using the object form of the ngOptions syntax.  Without this the second test would fail.